### PR TITLE
Fix incomplete pull by git-bug bridge

### PR DESCRIPTION
**Pull request by rng-dynamics**
Work in progress. The github bridge contains a bug documented in issue #385.
This commit shows what is the problem. There might be more problems. The
previous GraphQL query did not only retrieve new timeline items, but additionally
it acted on the issues. I have changed the query for timeline items and there are
much less wrong imports now. (Are there any malformed imports left?) There
might be similar problems with user-content-edit queries. I would like to rework
the entire bridge/github/iterator.go file in the near future in order to create a
reliable fix for this bug.

### CODE REVIEW

**Comment by rng-dynamics** [comment]
Note: this PR is a proof of concept.

**Comment by rng-dynamics** [comment]
@MichaelMure: I am wondering if tests with a mock using httptest would be worth the effort. In my opinion tests which are independent of the actual github API would be very valuable. I just don't have experience with testing a web API client. Do you think it is possible to mock the GraphQL API with reasonable effort? What do you think?

**Comment by MichaelMure** [comment]
Thanks for looking into that.

Reading the code doesn't help me much understand the issue but if you have a good idea of what it is, at least that's progress :)

> but additionally it acted on the issues

What do you mean by that?

> I would like to rework the entire bridge/github/iterator.go file in the near future in order to create a reliable fix for this bug.

Be my guest. FYI, the bridge initially was a monolith with import/export and data retrieval fused together, which made it fairly unreadable. I came up with this iterator structure to decouple those so now it's a bit better but I'll certainly admit it could be better. Curious what you have in mind.

> I am wondering if tests with a mock using httptest would be worth the effort. In my opinion tests which are independent of the actual github API would be very valuable.

I'm down on having more tests so if you feel like it make sense, go for it. A key thing the current tests achieve though is to monitor breakage against the real API that we don't control. I would assume the Github API being sort of stable (even though it does change) but it's a different game with other bridges (see https://github.com/MichaelMure/git-bug/blob/master/bridge/gitlab/export_test.go#L245-L248). It's a bit annoying because you need credentials and need to make actual requests, but it's useful imho.


Note: as you are working on your own fork, github actions doesn't get access to the repo secrets when running the tests, which means bridges are not tested in this PR. I added you as a collaborator of the repo so if you push your branch here that should be fixed.

**Comment by rng-dynamics** [comment]
> >but additionally it acted on the issues
> 
> What do you mean by that?

I just tried to express that the bug is related to the handling of issue items. To be more specific: at the moment when git-bug needs to query more timeline items, then it uses the pagination data which refers to the next issue already.

> > I would like to rework the entire bridge/github/iterator.go file in the near future in order to create a reliable fix for this bug.
> 
> Be my guest. FYI, the bridge initially was a monolith with import/export and data retrieval fused together, which made it fairly unreadable. I came up with this iterator structure to decouple those so now it's a bit better but I'll certainly admit it could be better. Curious what you have in mind.

Yes, that's not a problem. I also did not intend to disapprove of the code. I am just trying to see what I can do and what I can improve.

At the moment for each issue there is one GraphQL query which fetches a list of timeline items and user content edits. If there exist more timeline items of user content edits, then git-bug queries for more. The code could get much simpler if we (1) first query a list of issues only (that is, multiple issues at once) and (2) for each issue query the timeline items and user content edits. This will increase the number of queries by (total-num-issues / query-capacity), which in my opinion should not be a problem. By simplifying the code running queries concurrently should become easy and we might get some performance improvements. I would keep the iterators function signatures the same but I would probably change most of the GraphQL queries and the interanals of the iterators accordingly.

>I'm down on having more tests so if you feel like it make sense, go for it. A key thing the current tests achieve though is to monitor breakage against the real API that we don't control. I would assume the Github API being sort of stable (even though it does change) but it's a different game with other bridges (see https://github.com/MichaelMure/git-bug/blob/master/bridge/gitlab/export_test.go#L245-L248). It's a bit annoying because you need credentials and need to make actual requests, but it's useful imho.

You are right! I could just try to provide some additional tests.

**Comment by MichaelMure** [comment]
> I also did not intend to disapprove of the code.

I did not intend to be sarcastic or something. You can disapprove the code, that's how something better is built.

> This will increase the number of queries by (total-num-issues / query-capacity), which in my opinion should not be a problem.

Yeah, the rate-limiting for GraphQL is funny. It's not a matter of how many requests, it's pretty about much how much data you pull. Speaking of rate-limiting, the github bridge doesn't handle that and fail if it trigger: https://github.com/MichaelMure/git-bug/issues/315. If you happen to trigger it, maybe look at the response to see if we can match something and introduce back-off?

**Comment by rng-dynamics** [comment]
>  #315

Good to know.

**Comment by rng-dynamics** [comment]
The import works now.
Well ..., I have tried to write some useful tests, but all of them have been brittle and I did not add them to my branch.
That's what I have for now. (I might commit some improvements to the github-bridge in a future pull request.)

File: bridge/github/iterator.go [hunk 1]
```diff
@@ -4,290 +4,228 @@ import (
...
 	"github.com/shurcooL/githubv4"
 )
 
-type indexer struct{ index int }
-
-type issueEditIterator struct {
-	index     int
-	query     issueEditQuery
-	variables map[string]interface{}
+type iterator struct {
```
Comment (MichaelMure) [inline bridge/github/iterator.go:11]: Do you mind copying the comments from the previous version if that's still valid?

File: bridge/github/iterator.go [hunk 2]
```diff
@@ -300,102 +238,106 @@ func (i *iterator) nextValidIssueEdit() bool {
...
+	}
+	tli.index = 0
+	return true
+}
 
-	commentEdits := i.commentEdit.query.Repository.Issues.Nodes[0].Timeline.Nodes[0].IssueComment.UserContentEdits
-	// this is not supposed to happen
-	if len(commentEdits.Nodes) == 0 {
-		i.timeline.commentEdit.index = -1
+func (i *iterator) NextCommentEdit() bool {
```
Comment (MichaelMure) [inline bridge/github/iterator.go:315]: Same here and I suppose in others places.

**Review by MichaelMure — COMMENTED** [review]
Thanks!

It's a bit PR so I'll have a more thorough look later, just giving some obvious comments.

Any idea what made your test flacky? How flacky was that?

**Comment by rng-dynamics** [comment]
> Note: as you are working on your own fork, github actions doesn't get access to the repo secrets when running the tests, which means bridges are not tested in this PR. I added you as a collaborator of the repo so if you push your branch here that should be fixed.

Do the test work now? I still see

`=== RUN   Test_Importer
Building identity cache... Done.
Building bug cache... Done.
    import_test.go:141: Env var GITHUB_TOKEN_PRIVATE missing
--- SKIP: Test_Importer (0.01s)`

**Comment by bqv** [comment]
> might commit some improvements to the github-bridge in a future pull request

Please do! I'd love to use this more but the ratelimit is a dealbreaker right now :(

**Comment by bqv** [comment]
I attempted to test this, I'm seeing:

```
import error: Parse error on "{" (LCURLY) at [1, 234]
```

**Comment by rng-dynamics** [comment]
> I attempted to test this, I'm seeing:
> 
> ```
> import error: Parse error on "{" (LCURLY) at [1, 234]
> ```
From which repository are you importing?

**Comment by bqv** [comment]
I tried multiple, same error on all!

**Comment by rng-dynamics** [comment]
> Any idea what made your test flacky? How flacky was that?

I was experimenting with mocking the Github API. I would return hard-coded graphql responses to the shurcooL/githubv4 library. But the mocks are not simple and, I believe, hard to to maintain. Additionally, this approach is strongly coupled to the graphql queries (the type definitions in bridge/github/import_query.go). As a consequence, every time one changes the queries, then one also has to change the hard to understand mocks in the tests. When I looked at that, I thought the tests which you have now (calling the real github API) are actually better.

I am still experimenting with it. I am wondering if the github API needs to be kept out of the tests all together. The implementation of the iterator would probably get a bit less straight-forward. (E.g., I would like to try to internally split the NextIssue() call into functions nextIssueStep1(), maybeQueryIssue(), and nextIssueStep3() in such a way that maybeQueryIssue() is as simple as possible. Then we should be able to test the two functions nextIssueStep1() and nextIssueStep3() without coupling to a lot of implementation details.)

In any case, I would welcome suggestions.

File: bridge/github/iterator.go (hunk 1 above)
Comment (rng-dynamics) [reply bridge/github/iterator.go:11]: Sure, I will add them soon!

**Comment by MichaelMure** [comment]
> I thought I had accepted your invitation as a collaborator.
> It would just be nice to see them pass.

I believe you also need to push your branch to this repo.

**Comment by MichaelMure** [comment]
You could store the response from github and make a mock that return those. It would reduce to a simple-ish `http.Handler`. But then we wouldn't test that the bridge *keep working*, that is, if github didn't change their API or something...

**Comment by rng-dynamics** [comment]
Closed in favour of #534
(Switching the source repository)

**Pull request closed without merging**